pkg/multicast/multicast.go

```diff
@@ -411,149 +420,139 @@ func (c *Client) Handle(m *sbe.SbeGoMarshaller, r io.Reader) error {
+	q := make([]Bytes, 0)
 
-		case EventTypeTicker:
-			ticker := event.Data.(models.TickerNotification)
-			c.Emit(newTickerNotificationChannel(ticker.InstrumentName), &event.Data)
-		}
+	return &Pool{
+		mu:             &sync.RWMutex{},
+		queue:          q,
+		maxPackageSize: maxPackageSize,
 	}
-
-	return nil
 }
 
-func (c *Client) getConnection(addr string) *net.UDPConn {
-	c.mu.Lock()
-	defer c.mu.Unlock()
+func (p *Pool) Get() Bytes {
+	p.mu.Lock()
+	defer p.mu.Unlock()
 
-	return c.addrConnMap[addr]
+	if length := len(p.queue); length > 0 {
+		item := p.queue[length-1]
+		p.queue[length-1] = nil
+		p.queue = p.queue[:length-1]
+		return item
+	}
+	return make(Bytes, p.maxPackageSize)
 }
 
-func (c *Client) setConnection(addr string, conn *net.UDPConn) {
-	c.mu.Lock()
-	defer c.mu.Unlock()
+func (p *Pool) Put(bs Bytes) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
 
-	c.addrConnMap[addr] = conn
+	p.queue = append(p.queue, bs)
 }
 
-func (c *Client) delConnection(addr string) {
-	c.mu.Lock()
-	defer c.mu.Unlock()
-
-	delete(c.addrConnMap, addr)
-}
+// ListenToEvents listens to a list of udp addresses on given network interface.
+func (c *Client) ListenToEvents(ctx context.Context) error {
+	packetConn, err := net.ListenPacket("upd4", fmt.Sprintf("0.0.0.0:%d", c.port))
 
-func (c *Client) listenToMulticastUDP(addr string) (*net.UDPConn, error) {
-	udpAddr, err := net.ResolveUDPAddr("udp", addr)
 	if err != nil {
-		c.log.Errorw("failed to resolve UDP address", "err", err)
-		return nil, err
+		c.log.Errorw("failed to initiate packet connection", "err", err, "port", c.port)
+		return err
 	}
+	defer packetConn.Close()
 
-	udpConn, err := net.ListenMulticastUDP("udp", c.inf, udpAddr)
-	if err != nil {
-		c.log.Errorw("failed to listen to multicast UDP", "error", err)
-		return nil, err
-	}
+	ipv4PacketConn := ipv4.NewPacketConn(packetConn)
 
-	err = udpConn.SetReadBuffer(defaultReadBufferSize)
+	ipGroups := make([]net.IP, len(c.ipAddrs))
+
+	err = ipv4PacketConn.SetControlMessage(ipv4.FlagDst, true)
 	if err != nil {
-		c.log.Errorw("fail to set read buffer for UDP", "error", err)
-		return nil, err
+		c.log.Errorw("failed to set control message", "err", err)
 	}
 
-	c.setConnection(addr, udpConn)
-	return udpConn, nil
-}
+	for _, ipAddr := range c.ipAddrs {
+		group := net.ParseIP(ipAddr)
+		if group == nil {
+			return ErrInvalidIpv4Address
+		}
 
-// ListenToEventsForAddress listens to one udp address on given network interface.
-func (c *Client) ListenToEventsForAddress(ctx context.Context, addr string) error {
-	l := c.log.With("addr", addr)
-	dataCh := make(chan []byte, defaultDataChSize)
-	udpConn, err := c.listenToMulticastUDP(addr)
-	if err != nil {
-		l.Errorw("failed to listen to multicast UDP", "err", err)
-		return err
+		err := ipv4PacketConn.JoinGroup(c.inf, &net.UDPAddr{IP: group})
+		if err != nil {
+			c.log.Errorw("failed to join group", "group", group, "err", err, "ipAddr", ipAddr)
+		}
+
+		ipGroups = append(ipGroups, group)
 	}
```

Comment:
All of the above code should move to a function for setting up a connection for multiple multicast groups.